### PR TITLE
discoverd: Die sdutil!

### PR DESCRIPTION
**Pull request by archSeer**
This is ready for merge. I'm writing a quick prototype of how the CLI-side interactions with these services would go.

### CODE REVIEW

File: receiver/flynn-receive.go [hunk 1]
```diff
@@ -103,11 +104,15 @@ func main() {
...
-				proc.Env = make(map[string]string)
-			}
-			proc.Env["SD_NAME"] = app.Name + "-web"
+			proc.Ports = []ct.Port{{
+				Port:  8080,
+				Proto: "tcp",
+				Service: &host.PortService{
+					Name:   app.Name + "-web",
+					Create: true,
+					Check:  &host.PortServiceCheck{Type: "http"},
```
Comment (archSeer) [inline receiver/flynn-receive.go]: HTTP check by default. Works for majority of cases, but might not fit where the user has a resource that is not available under `/`, however I think we can just note that in the usage manual, and then have instructions on how to change the service checker using the CLI.

File: host/containerinit/init.go [hunk 2]
```diff
@@ -395,6 +401,71 @@ func getCmdPath(c *Config) (string, error) {
...
+		}
+	}
+	portStr := strconv.Itoa(port.Port)
+	if portStr == "0" {
+		portStr = env["PORT"]
+	}
+	localAddr := fmt.Sprintf("127.0.0.1:%s", portStr)
+	addr := fmt.Sprintf("%s:%s", env["EXTERNAL_IP"], portStr)
+	var check health.Check
+	switch config.Check.Type {
```
Comment (archSeer) [inline host/containerinit/init.go]: Should probably add support for registering a service without a checker? Would skip the `health.Registration` and instead just return a heartbeater which gets closed when the process exits.

**Comment by archSeer** [comment]
@titanous I'm thinking of renaming the types `PortService` -> `Service`, `PortServiceChecker` -> `Checker/HealthChecker`. WDYT?

File: receiver/flynn-receive.go (hunk 1 above)
Comment (titanous) [inline receiver/flynn-receive.go]: We could add a loose check that accepts any status code (maybe status code `-1`), what do you think?

File: host/containerinit/init.go (hunk 2 above)
Comment (titanous) [inline host/containerinit/init.go]: Yes, absolutely.

File: host/containerinit/init.go [hunk 3]
```diff
@@ -535,9 +619,6 @@ func Main() {
 		log.Fatalf("Unable to unmarshal config: %v", err)
 	}
 
-	// Propagate the plugin-specific container env variable
-	config.Env["container"] = os.Getenv("container")
-
```
Comment (titanous) [inline host/containerinit/init.go]: Why was this removed?

**Comment by titanous** [comment]
Changing the type names is fine, `HealthChecker` sounds good.

File: host/containerinit/init.go [hunk 4]
```diff
@@ -395,6 +401,71 @@ func getCmdPath(c *Config) (string, error) {
...
+		// TODO: maybe reuse maybeAddService() from the client
+		if err := client.AddService(config.Name); err != nil {
+			if je, ok := err.(hh.JSONError); !ok || je.Code != hh.ObjectExistsError {
+				return nil, fmt.Errorf("something went wrong with discoverd: %s", err)
+			}
+		}
+	}
+	portStr := strconv.Itoa(port.Port)
+	if portStr == "0" {
+		portStr = env["PORT"]
```
Comment (titanous) [inline host/containerinit/init.go]: Do this before converting to a string.

File: host/containerinit/init.go [hunk 5]
```diff
@@ -395,6 +401,71 @@ func getCmdPath(c *Config) (string, error) {
...
+	}
+	localAddr := fmt.Sprintf("127.0.0.1:%s", portStr)
+	addr := fmt.Sprintf("%s:%s", env["EXTERNAL_IP"], portStr)
```
Comment (titanous) [inline host/containerinit/init.go]: No need to convert  port to a string if you are using `fmt`.

File: host/containerinit/init.go [hunk 6]
```diff
@@ -395,6 +401,71 @@ func getCmdPath(c *Config) (string, error) {
...
+	if config.Check.KillDown {
+		// ignore events for the first StartTimeout interval
+		events := make(chan health.MonitorEvent)
+		reg.Events = events
+		go func() {
+			for e := range events {
+				if e.Status != health.MonitorStatusDown {
+					continue
+				}
+				// kill process
```
Comment (titanous) [inline host/containerinit/init.go]: Is this a TODO?

File: host/containerinit/init.go [hunk 7]
```diff
@@ -513,6 +584,19 @@ func containerInitApp(c *Config) error {
...
 
 	init.mtx.Unlock() // Allow calls
+	// monitor services
+	for _, port := range c.Ports {
+		if port.Service == nil {
+			continue
+		}
+		hb, err := monitor(port, c.Env)
+		if err != nil {
+			// TODO: handle err
```
Comment (titanous) [inline host/containerinit/init.go]: TODO needs handling?

**Comment by titanous** [comment]
This should have a bunch of integration tests covering each option of the service check, especially stuff like `KillDown` and thresholds.

File: host/containerinit/init.go (hunk 3 above)
Comment (archSeer) [inline host/containerinit/init.go]: Is it used anywhere? I think it's a leftover from the original code.

File: host/containerinit/init.go (hunk 4 above)
Comment (archSeer) [inline host/containerinit/init.go]: `env["PORT"]` is a string, so I need to convert that to an integer then.

File: host/containerinit/init.go (hunk 6 above)
Comment (archSeer) [inline host/containerinit/init.go]: Ah, yes.

File: host/containerinit/init.go (hunk 7 above)
Comment (archSeer) [inline host/containerinit/init.go]: I am unsure how to handle this. Should the entire application fail, if the Register fails to spin up for some reason, or should we simply skip it (like we do now) and let the application run?

Comment (titanous) [inline host/containerinit/init.go]: Entire app should fail.

File: host/containerinit/init.go (hunk 3 above)
Comment (titanous) [inline host/containerinit/init.go]: It sets an environment variable that should be passed on. It is 'used' in the sense that that env var should be available.

File: host/containerinit/init.go (hunk 4 above)
Comment (titanous) [inline host/containerinit/init.go]: Ah, misunderstood this. Can't we just serialize the info after port assignment?

File: host/containerinit/init.go (hunk 3 above)
Comment (archSeer) [inline host/containerinit/init.go]: Yes, I figured that `os.Getenv("container") == ""`, since I couldn't find anything setting `container`

Comment (titanous) [inline host/containerinit/init.go]: libvirt-lxc sets it. Check out the environment of a container.

File: host/containerinit/init.go (hunk 4 above)
Comment (archSeer) [inline host/containerinit/init.go]: What do you mean by that?

File: receiver/flynn-receive.go (hunk 1 above)
Comment (archSeer) [inline receiver/flynn-receive.go]: I think that could work. We could also just use the TCP checker?

Comment (titanous) [inline receiver/flynn-receive.go]: Yeah, probably easier to just do TCP and won't show up in the logs, we can always change it later.

**Comment by archSeer** [comment]
@titanous Do we need to test all cases though? I'd test the `KillDown + StartTimeout`, but all of the `Monitor` and `Checker` options are already tested in the discoverd unit tests.

**Comment by titanous** [comment]
A sanity check or two wouldn't hurt, but you don't need to go overboard.

**Comment by archSeer** [comment]
@titanous Tests pass.

File: host/containerinit/init.go [hunk 8]
```diff
@@ -395,6 +400,89 @@ func getCmdPath(c *Config) (string, error) {
...
+
+	if config.Create {
+		// TODO: maybe reuse maybeAddService() from the client
+		if err := client.AddService(config.Name); err != nil {
+			if je, ok := err.(hh.JSONError); !ok || je.Code != hh.ObjectExistsError {
+				return nil, fmt.Errorf("something went wrong with discoverd: %s", err)
+			}
+		}
+	}
+	localAddr := fmt.Sprintf("127.0.0.1:%v", port.Port)
```
Comment (titanous) [inline host/containerinit/init.go]: Do we need this? Can't we just use `addr`?

File: host/containerinit/init.go [hunk 9]
```diff
@@ -395,6 +400,89 @@ func getCmdPath(c *Config) (string, error) {
...
+			Threshold: config.Check.Threshold,
+		}.Run,
+		Check: check,
+	}
+
+	if config.Check.KillDown {
+		reg.Events = make(chan health.MonitorEvent)
+		go func() {
+			if config.Check.StartTimeout == 0 {
+				config.Check.StartTimeout = 10
```
Comment (titanous) [inline host/containerinit/init.go]: This would be 10 nanoseconds, which is probably not what we want.

File: host/containerinit/init.go [hunk 10]
```diff
@@ -395,6 +400,89 @@ func getCmdPath(c *Config) (string, error) {
...
+			}
+
+			start := false
+			lastStatus := health.MonitorStatusDown
+
+			maybeKill := func() {
+				if lastStatus == health.MonitorStatusDown {
```
Comment (titanous) [inline host/containerinit/init.go]: This check is racy

File: host/containerinit/init.go [hunk 11]
```diff
@@ -395,6 +400,89 @@ func getCmdPath(c *Config) (string, error) {
...
+					container.Signal(int(syscall.SIGKILL), &struct{}{})
+				}
+			}
+			go func() {
+				// ignore events for the first StartTimeout interval
+				<-time.After(config.Check.StartTimeout * time.Second)
+				maybeKill() // check if the app is down
+				start = true
```
Comment (titanous) [inline host/containerinit/init.go]: This set is racy

File: host/containerinit/init.go (hunk 10 above)
Comment (archSeer) [inline host/containerinit/init.go]: Should I just wrap all of it with a `sync.RWMutex`, or do you have a better approach in mind?

File: host/containerinit/init.go (hunk 9 above)
Comment (archSeer) [inline host/containerinit/init.go]: Hmm, I'll change it for clarity, but the test definitely takes 10 seconds to timeout (passing tests take 2s, failing test takes 12s)

File: host/containerinit/init.go (hunk 8 above)
Comment (archSeer) [inline host/containerinit/init.go]: I ran into problems originally, but I think that might have happened because of other problems. Now that we have tests, I'll give it a try.

File: host/containerinit/init.go (hunk 10 above)
Comment (titanous) [inline host/containerinit/init.go]: yeah, a `sync.Mutex` is fine, probably overkill to use `RWMutex`

**Comment by archSeer** [comment]
@titanous Comments addressed.

File: host/containerinit/init.go [hunk 12]
```diff
@@ -395,6 +400,94 @@ func getCmdPath(c *Config) (string, error) {
...
+	if config.Check.KillDown {
+		reg.Events = make(chan health.MonitorEvent)
+		go func() {
+			if config.Check.StartTimeout == 0 {
+				config.Check.StartTimeout = 10 * time.Second
+			}
+
+			start := false
+			lastStatus := health.MonitorStatusDown
+			mtx := &sync.Mutex{}
```
Comment (titanous) [inline host/containerinit/init.go]: Can just do `var mtx sync.Mutex`

Comment (archSeer) [inline host/containerinit/init.go]: Addressed.

File: test/apps/ping/main.go [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+
+const service = "ping-service"
+
+func main() {
+	defer shutdown.Exit()
+
+	port := os.Getenv("PORT")
+	addr := ":" + port
+
+	hb, err := discoverd.AddServiceAndRegister(service, addr)
```
Comment (titanous) [inline test/apps/ping/main.go]: Why does this service register with discoverd?

Comment (archSeer) [inline test/apps/ping/main.go]: Ouch, totally missed that, I think I copied the boilerplate code from the TCP app...

File: test/test_healthcheck.go [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+
+func (s *HealthcheckSuite) TestWithoutChecker(t *c.C) {
+	// start app with a service but no checker
+	app, _ := s.createAppWithService(t, "ping", &host.Service{
+		Name:   "ping-without-checker",
+		Create: true,
+	})
+	t.Assert(flynn(t, "/", "-a", app.Name, "scale", "ping=1"), Succeeds)
+	// make sure app is registered and unregistered when the process terminates
+	_, err := s.discoverdClient(t).Instances("ping-without-checker", 10*time.Second)
```
Comment (titanous) [inline test/test_healthcheck.go]: This doesn't do what it says on the tin. Where is the verification that it is unregistered?

File: test/test_healthcheck.go [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+	c "github.com/flynn/flynn/Godeps/_workspace/src/github.com/flynn/go-check"
+	ct "github.com/flynn/flynn/controller/types"
+	"github.com/flynn/flynn/host/types"
+)
+
+type HealthcheckSuite struct {
+	Helper
+}
+
+var _ = c.Suite(&HealthcheckSuite{})
```
Comment (titanous) [inline test/test_healthcheck.go]: This should be a ConcurrentSuite?

File: test/test_healthcheck.go [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+		Create: true,
+		Check:  &host.HealthCheck{Type: "tcp"},
+	})
+	t.Assert(flynn(t, "/", "-a", app.Name, "scale", "printer=1"), Succeeds)
+	// confirm that it's never registered
+	_, err := s.discoverdClient(t).Instances("healthcheck-failure", 10*time.Second)
+	t.Assert(err, c.NotNil)
+}
+
+func (s *HealthcheckSuite) TestKilldown(t *c.C) {
```
Comment (titanous) [inline test/test_healthcheck.go]: s/Killdown/KillDown/

File: test/test_healthcheck.go [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+	// start an app that is failing checks /w killdown
+	app, _ := s.createAppWithService(t, "printer", &host.Service{
+		Name:   "healthcheck-killdown",
+		Create: true,
+		Check:  &host.HealthCheck{Type: "tcp", KillDown: true},
```
Comment (titanous) [inline test/test_healthcheck.go]: This should drop the `StartTimeout`

File: test/test_healthcheck.go [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+	// start an app that is failing checks
+	app, _ := s.createAppWithService(t, "printer", &host.Service{
+		Name:   "healthcheck-failure",
+		Create: true,
+		Check:  &host.HealthCheck{Type: "tcp"},
+	})
+	t.Assert(flynn(t, "/", "-a", app.Name, "scale", "printer=1"), Succeeds)
+	// confirm that it's never registered
+	_, err := s.discoverdClient(t).Instances("healthcheck-failure", 10*time.Second)
+	t.Assert(err, c.NotNil)
```
Comment (titanous) [inline test/test_healthcheck.go]: Drop the StartTimeout so we don't have to wait ten seconds for this test to fail.

**Comment by archSeer** [comment]
@titanous Addressed again. Note that TestFailure takes 11s for some reason, the `current == nil` event is sent straight away, but the down event is delayed. Might have something to do with `Threshold` on the monitor?

**Comment by titanous** [comment]
@archSeer Yes, sounds like a threshold/timeout issue.

**Comment by archSeer** [comment]
@titanous Ready for another pass, should pass CI after the server is restarted.

File: discoverd/health/register_test.go [hunk 19]
```diff
@@ -136,7 +136,11 @@ func (RegisterSuite) TestRegister(c *C) {
 			Events:    make(chan MonitorEvent),
 		}
 		hb := reg.Register()
-		defer hb.Close()
+		defer func() {
+			go func() { <-unregisterChan }()
```
Comment (titanous) [inline discoverd/health/register_test.go]: What does this goroutine do?

Comment (archSeer) [inline discoverd/health/register_test.go]: 1559e5fc620367c9e1c1c1d47c03420598feb40f calls the wrapped heartbeater's close method, which then [triggers the mocked `Close()` method](https://github.com/flynn/flynn/blob/1559e5fc620367c9e1c1c1d47c03420598feb40f/discoverd/health/register_test.go#L61), that sends a notification to the `unregisterChan`. I'm making sure the channel gets drained here, so that the function does not block indefinitely.

**Comment by titanous** [comment]
LGTM
